File: examples/android/src/main/java/com/dropbox/core/examples/android/UploadFileTask.java
package com.dropbox.core.examples.android;

import android.content.Context;
import android.net.Uri;
import android.os.AsyncTask;

import com.dropbox.core.DbxException;
import com.dropbox.core.v2.DbxFiles;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * Async task to upload a file to a directory
 */
class UploadFileTask extends AsyncTask<String, Void, DbxFiles.FileMetadata> {

    private final Context mContext;
    private final DbxFiles mFilesClient;
    private Exception mException;
    private Callback mCallback;

    public interface Callback {
        void onUploadComplete(DbxFiles.FileMetadata result);
        void onError(Exception e);
    }

    UploadFileTask(Context context, DbxFiles filesClient, Callback callback) {
        mContext = context;
        mFilesClient = filesClient;
        mCallback = callback;
    }

    @Override
    protected void onPostExecute(DbxFiles.FileMetadata result) {
        super.onPostExecute(result);
        if (mException != null) {
            mCallback.onError(mException);
        } else if (result == null) {
            mCallback.onError(null);
        } else {
            mCallback.onUploadComplete(result);
        }
    }

    @Override
    protected DbxFiles.FileMetadata doInBackground(String... params) {
        String localUri = params[0];
        File localFile = UriHelpers.getFileForUri(mContext, Uri.parse(localUri));

        if (localFile != null) {
            String remoteFolderPath = params[1];

            // Note - this is not ensuring the name is a valid dropbox file name
            String remoteFileName = localFile.getName();

            try {
                InputStream inputStream = new FileInputStream(localFile);
                try {
                    mFilesClient.uploadBuilder(remoteFolderPath + "/" + remoteFileName)
                            .mode(DbxFiles.WriteMode.overwrite)
                            .run(inputStream);
                } finally {
                    inputStream.close();
                }
            } catch (DbxException | IOException e) {
                e.printStackTrace();
                mException = e;
            }
        }

        return null;
    }
}

File: examples/android/src/main/java/com/dropbox/core/examples/android/UploadFileTask.kt
package com.dropbox.core.examples.android

import android.content.Context
import android.net.Uri
import com.dropbox.core.DbxException
import com.dropbox.core.v2.DbxClientV2
import com.dropbox.core.v2.files.FileMetadata
import com.dropbox.core.v2.files.WriteMode
import java.io.FileInputStream
import java.io.IOException
import kotlin.coroutines.CoroutineContext
import kotlinx.coroutines.withContext


sealed interface FileMetadataApiResult {
    class Success(val fileMetadata: FileMetadata) : FileMetadataApiResult
    class Error(val e: Exception) : FileMetadataApiResult
}

/**
 * Async task to upload a file to a directory
 */
internal class UploadFileTask(
    private val mContext: Context,
    private val mDbxClient: DbxClientV2,
    private val dispatcher: CoroutineContext,
) {
    suspend fun execute(localUri: String, remoteFolderPath: String): FileMetadataApiResult =
        withContext(dispatcher) {
            val localFile = UriHelpers.getFileForUri(mContext, Uri.parse(localUri))
            if (localFile != null) {
                // Note - this is not ensuring the name is a valid dropbox file name
                val remoteFileName = localFile.name
                try {
                    FileInputStream(localFile).use { inputStream ->
                        val metadata = mDbxClient.files().uploadBuilder(
                            "$remoteFolderPath/$remoteFileName"
                        )
                            .withMode(WriteMode.OVERWRITE)
                            .uploadAndFinish(inputStream)
                        FileMetadataApiResult.Success(metadata)
                    }
                } catch (e: DbxException) {
                    FileMetadataApiResult.Error(e)
                } catch (e: IOException) {
                    FileMetadataApiResult.Error(e)
                }
            } else {
                FileMetadataApiResult.Error(Exception("Local File to Upload Not Found"))
            }
        }
}